compiler/natives/os/os.go

```diff
@@ -24,3 +24,27 @@ func init() {
 }
 
 func runtime_beforeExit() {}
+
+// Override the builtin os.Hostname(). The builtin Hostname() will result in
+// a non-working syscall on most (all?) platforms
+func Hostname() (string, error) {
+	// Browser
+	// The browser doesn't know anything about any hostname, so we use the host part of
+	// the current url.
+	if location := js.Global.Get("location"); location != js.Undefined {
+		if browserHostname := location.Get("hostname"); browserHostname != js.Undefined {
+			return browserHostname.String(), nil
+		}
+	}
+
+	// Node.js
+	// We call node's require("os").hostname() which should return current hostname
+	if require := js.Global.Get("require"); require != js.Undefined {
+		// "os" is in nodejs core, it should always exist
+		if nodeHostname := require.Invoke("os").Call("hostname"); nodeHostname != js.Undefined {
```

Comment:
For that to happen a few planets have to be aligned:

- The browser doesn't implement the global `location` object. Is there any browsers that doesn't?
- The user uses browserify.
- The user expects `os.Hostname()` to not fail.

After this PR the result would be a catastrophic Javascript failure.
Before this PR the result would be a catastrophic gopherjs failure.

I think we can safely ignore this corner case, but I'm open for suggestions.